File: web-service/src/services/paste.service.ts
import type CreatePasteResponse from '@/types/payload/paste/create.res'
import type CreatePasteRequest from '@/types/payload/paste/create.req'
import { ApiUtil } from '@/utils/api.util'
import type GetOwnPastesRequest from '@/types/payload/paste/get-own.req'
import type CommonPaginationResponse from '@/types/payload/common-pagination.res'
import type PasteResponse from '@/types/payload/paste/paste.res'

interface IPasteService {
  create: (request: CreatePasteRequest) => Promise<CreatePasteResponse>,
  getOwnPastes: (params: GetOwnPastesRequest) => Promise<CommonPaginationResponse<PasteResponse>>
  getOwnPasteById: (pasteId: string) => Promise<PasteResponse>,
}

const PasteService: IPasteService = {
  create: (request): Promise<CreatePasteResponse> => {
    return new Promise((resolve, reject) => {
      ApiUtil.post('/pastes', request)
        .then((response) => {
          resolve(response)
        })
        .catch((error) => {
          reject(error)
        })
    })
  },

  getOwnPastes: (params): Promise<CommonPaginationResponse<PasteResponse>> => {
    return new Promise((resolve, reject) => {
      ApiUtil.get('/pastes/own', { params })
        .then((response) => {
          resolve(response)
        })
        .catch((error) => {
          reject(error)
        })
    })
  },

  getOwnPasteById: (pasteId: string): Promise<PasteResponse> => {
    return new Promise((resolve, reject) => {
      ApiUtil.get(`/pastes/${pasteId}`)
        .then((response) => {
          resolve(response)
        })
        .catch((error) => {
          reject(error)
        })
    })
  }
}

export default PasteService

File: paste-service/src/main/java/dev/hieplp/pastebin/paste/service/PasteService.java
package dev.hieplp.pastebin.paste.service;

import dev.hieplp.pastebin.common.payload.response.CommonPaginationResponse;
import dev.hieplp.pastebin.paste.payload.request.CreatePasteRequest;
import dev.hieplp.pastebin.paste.payload.request.GetOwnPastesRequest;
import dev.hieplp.pastebin.paste.payload.request.UpdatePasteRequest;
import dev.hieplp.pastebin.paste.payload.response.CreatePasteResponse;
import dev.hieplp.pastebin.paste.payload.response.PasteResponse;
import dev.hieplp.pastebin.paste.payload.response.UpdatePasteResponse;

public interface PasteService {
    /**
     * Create paste
     *
     * @param request Create paste request
     * @param ownerId Owner id
     * @return Create paste response
     * @throws dev.hieplp.pastebin.common.exception.DuplicateException If alias is already taken
     */
    CreatePasteResponse create(CreatePasteRequest request, String ownerId);

    /**
     * Update paste
     *
     * @param pasteId   Paste id
     * @param request   Update paste request
     * @param updatedBy ID of user who updated the paste
     * @return Update paste response
     * @throws dev.hieplp.pastebin.common.exception.AccessDeniedException If user is not the owner of paste
     * @throws dev.hieplp.pastebin.common.exception.NotFoundException     If paste not found
     * @throws dev.hieplp.pastebin.common.exception.DuplicateException    If alias is already taken
     */
    UpdatePasteResponse update(String pasteId, UpdatePasteRequest request, String updatedBy);

    /**
     * Delete paste
     *
     * @param pasteId   Paste id
     * @param deletedBy ID of user who deleted the paste
     * @throws dev.hieplp.pastebin.common.exception.NotFoundException     If paste not found
     * @throws dev.hieplp.pastebin.common.exception.AccessDeniedException If user is not the owner of paste
     */
    void delete(String pasteId, String deletedBy);

    /**
     * Get paste
     *
     * @param pasteId Paste id
     * @param userId  User id
     * @return Paste response
     * @throws dev.hieplp.pastebin.common.exception.NotFoundException     If paste not found
     * @throws dev.hieplp.pastebin.common.exception.AccessDeniedException If user is not the owner of paste
     */
    PasteResponse get(String pasteId, String userId);

    /**
     * Get public paste by alias
     *
     * @param username Username
     * @param alias    Alias
     * @return Paste response
     */
    PasteResponse getByUsernameAndAlias(String username, String alias);

    /**
     * Get own pastes
     *
     * @param request Get own pastes request
     * @param ownerId Owner id
     * @return Common pagination response of paste response
     */
    CommonPaginationResponse<PasteResponse> getOwnPastes(GetOwnPastesRequest request, String ownerId);
}
